### PR TITLE
Frida Refactor: Split FridaHelper into each Runtime

### PR DESCRIPTION
**Pull request by tokatoka**
This pr includes several frida refactoring
#307 
- [x] Split coverage runtime
- [x] Test builing for aarch64 
- [x] Clean up

### CODE REVIEW

**Comment by tokatoka** [comment]
I think I should split helper, and make a CoverageRuntime

File: libafl_frida/src/helper.rs [hunk 1]
```diff
@@ -471,6 +476,42 @@ impl<'a> FridaInstrumentationHelper<'a> {
         Aarch64Register::from_u32(regint as u32).unwrap()
     }
 
+    fn generate_maybe_log_blob(&mut self) -> Box<[u8]> {
+        macro_rules! maybe_log{
+            ($ops:ident) => {dynasm!($ops
+                ;   .arch x64
```
Comment (s1341) [inline libafl_frida/src/helper.rs:482]: We done my really need the macro, do we?

Comment (tokatoka) [reply libafl_frida/src/helper.rs:482]: No 😉 
I plan to move this to CoverageRuntime.

**Comment by s1341** [comment]
Yes splitting out CoverageRuntime is a good idea. It might also be worth moving the `is_interesting` functions out of helper into their various Runtimes.

**Comment by s1341** [comment]
Also it doesn’t look like you used your new function anywhere, did you? It should be called only once and the result cached.

**Comment by tokatoka** [comment]
no, I can't use it yet.
that's why it's still a draft.
At first I tried to call this from FridaHelper->new(),  but this does not look because we can't call `self.generate_maybe_log_blob` to slot the blob into self in `new()`. 
(what I can do is like calling `FridaInstrumentationHelper::<'a>::generate_maybe_log_blob(&mut helper)` but that doesn't look nice)
To achieve this, I find it's nicer to create a CoverageRuntime and we can call that generate-blob function in something like `CoverageRuntime->init()`

**Comment by s1341** [comment]
Cool. Makes sense. I’ll let you work. ;)

**Comment by tokatoka** [comment]
This change already works on amd64
i'm gonna test building for aarch64 tomorrow @s1341 
(I hope we have a android ci tbh...)

**Comment by tokatoka** [comment]
how about moving to `hook_func`s to other file like `asan_hooks`? @s1341 
currently it is just... toooooo long

**Comment by s1341** [comment]
Good idea. Do you want to do it in this PR or in another?

**Comment by tokatoka** [comment]
The fix would be simple. I would do it in this pr.

**Comment by s1341** [comment]
Ok. Go ahead.

**Comment by s1341** [comment]
Let me know when you want me to test on aarch64.

**Comment by tokatoka** [comment]
I come to feel that I should do these in seperate prs...

**Comment by tokatoka** [comment]
so that I can track down my mistakes if I broke something.

**Comment by tokatoka** [comment]
I checked building for aarch64.

**Comment by tokatoka** [comment]
I would like to get this merged and work on remaining issues in other prs

File: libafl_frida/src/coverage_rt.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+use frida_gum::instruction_writer::{Aarch64Register, IndexMode};
+
+use frida_gum::{instruction_writer::InstructionWriter, stalker::StalkerOutput};
+
+/// (Default) map size for frida coverage reporting
+pub const MAP_SIZE: usize = 64 * 1024;
+
+pub struct CoverageRuntime {
+    map: [u8; MAP_SIZE],
+    previous_pc: [u64; 1],
```
Comment (s1341) [inline libafl_frida/src/coverage_rt.rs:16]: Why is this an array?

File: libafl_frida/src/coverage_rt.rs [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+            current_log_impl: 0,
+            blob_maybe_log: None,
+        }
+    }
+
+    pub fn init(&mut self) {
+        self.generate_maybe_log_blob();
+    }
+
+    pub fn map_ptr(&mut self) -> *mut u8 {
```
Comment (s1341) [inline libafl_frida/src/coverage_rt.rs:42]: shouldn't this be `map_ptr_mut` to stick to the normal rust conventions?

File: libafl_frida/src/coverage_rt.rs [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        let mut ops = dynasmrt::VecAssembler::<dynasmrt::x64::X64Relocation>::new(0);
+        dynasm!(ops
+            ;   .arch aarch64
+            ;   stp x1, x2, [sp, -0x10]!
+            ;   stp x3, x4, [sp, -0x10]!
+            ;   ldr x1, >map_addr
+            ;   ldr x2, #0x38
+            ;   ldr x4, [x2]
+            ;   eor x4, x4, x0
+            ;   and x4, x4, 0xffff
```
Comment (s1341) [inline libafl_frida/src/coverage_rt.rs:64]: This should be `MAP_SIZE - 1` instead of `0xffff`

File: libafl_frida/src/coverage_rt.rs [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+    /// block.
+    #[cfg(target_arch = "aarch64")]
+    pub fn generate_maybe_log_blob(&mut self) {
+        let mut ops = dynasmrt::VecAssembler::<dynasmrt::x64::X64Relocation>::new(0);
+        dynasm!(ops
+            ;   .arch aarch64
+            ;   stp x1, x2, [sp, -0x10]!
+            ;   stp x3, x4, [sp, -0x10]!
+            ;   ldr x1, >map_addr
+            ;   ldr x2, #0x38
```
Comment (s1341) [inline libafl_frida/src/coverage_rt.rs:61]: This should be `>previous_loc`

**Review by s1341 — COMMENTED** [review]
A few nitpicks... but otherwise lgtm!

File: libafl_frida/src/coverage_rt.rs (hunk 2 above)
Comment (tokatoka) [reply libafl_frida/src/coverage_rt.rs:16]: it was an array before ;)
but i'll fix

Comment (s1341) [reply libafl_frida/src/coverage_rt.rs:16]: Ok. Didn't realize... Yeah if you can fix...

**Comment by tokatoka** [comment]
I fixed all, but didn't test how aarch64 build goes with the change.

**Comment by s1341** [comment]
I fixed a minor issue which was breaking aarch64 compilation, but running on aarch64 blows up for some reason. I'm debugging now.

**Comment by domenukk** [comment]
Can you maybe try to add the build to CI whole you do?

**Comment by s1341** [comment]
It requires adding/installing an ndk... which I don't know how to do under CI in a way that won't be very time-expensive.

**Comment by domenukk** [comment]
We already build parts of it, see
https://github.com/AFLplusplus/LibAFL/blob/9ab866336658d186e059432fbdaf8b7b16a8e605/.github/workflows/build_and_test.yml#L209

**Comment by s1341** [comment]
Hrm. I'll try to look into it.

**Comment by tokatoka** [comment]
> I fixed a minor issue which was breaking aarch64 compilation, but running on aarch64 blows up for some reason. I'm debugging now.

I observe that arm64 run is broken on main branch too.

**Comment by tokatoka** [comment]
mine fails at this
```
Thread 3.1 "frida_libpng" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0xfffff7c03430 (LWP 41363)]
0x0000aaaaaad30880 in nix::sys::mman::mmap (addr=0x400000000000, length=<optimized out>, prot=..., flags=..., fd=-1, offset=0)
    at /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/nix-0.23.0/src/sys/mman.rs:339
339         let ret = libc::mmap(addr, length, prot.bits(), flags.bits(), fd, offset);
```

but I'm not turning asan on.... 🤔

**Comment by s1341** [comment]
I think the coverage is broken. Haven't had a moment to debug today... Hopefully will get to it a little later.

**Comment by tokatoka** [comment]
I confirmed that this error already occurs back at f0daeb377e9df0fd664027d92d63178769d11986 this commit, (just before amd64 asan)

**Comment by s1341** [comment]
Is it possible that the machine you are running on doesn't have memory over-commit?

**Comment by tokatoka** [comment]
Oh, ... it is possible.
It is just a potato aws instance...
ok, so my comments would be unrelated, sorry.

**Comment by domenukk** [comment]
I don't think it's possible, they should be normal Linux OSses with normal configuration.

BTW @tokatoka in case you are spawning instances mainly for libafl dev, let us know, I'm sure we have some leftover gsoc funds for that or something :)

**Comment by tokatoka** [comment]
it's ok.
yesterday I used that for like 2 hours but just costed me 10 jpy (like less than 0.1 euro)

**Comment by andreafioraldi** [comment]
Should I merge?

**Comment by tokatoka** [comment]
no aarch64 run is broken, let's wait s1341

**Pull request merged**